pdns/recursordist/rec-zonetocache.cc

```diff
@@ -0,0 +1,288 @@
+  ComboAddress local = config.d_local;
+  if (local == ComboAddress()) {
+    local = pdns::getQueryLocalAddress(primary.sin4.sin_family, 0);
+  }
+
+  AXFRRetriever axfr(primary, d_zone, tt, &local, maxReceivedBytes, axfrTimeout);
+  Resolver::res_t nop;
+  vector<DNSRecord> chunk;
+  time_t axfrStart = time(nullptr);
+  time_t axfrNow = time(nullptr);
+
+  while (axfr.getChunk(nop, &chunk, (axfrStart + axfrTimeout - axfrNow))) {
+    for (auto& dr : chunk) {
+      parseDRForCache(dr);
+    }
+    axfrNow = time(nullptr);
+    if (axfrNow < axfrStart || axfrNow - axfrStart > axfrTimeout) {
+      throw PDNSException("Total AXFR time for zoneToCache exceeded!");
+    }
+  }
+}
+
+static std::vector<std::string> getLinesFromFile(const RecZoneToCache::Config& config)
+{
+
+  std::vector<std::string> lines;
+  std::ifstream stream(config.d_sources.at(0));
+  if (!stream) {
+    throw PDNSException("Cannot read file: " + config.d_sources.at(0));
+  }
+  std::string line;
+  while (std::getline(stream, line)) {
+    lines.push_back(line);
+  }
+  return lines;
+}
+
+static std::vector<std::string> getURL(const RecZoneToCache::Config& config)
+{
+  std::vector<std::string> lines;
+#ifdef HAVE_LIBCURL
+  MiniCurl mc;
+  ComboAddress local = config.d_local;
+  std::string reply = mc.getURL(config.d_sources.at(0), nullptr, local == ComboAddress() ? nullptr : &local, config.d_timeout, false, true);
+  if (config.d_maxReceivedBytes > 0 && reply.size() > config.d_maxReceivedBytes) {
+    // We should actually detect this *during* the GET
+    throw PDNSException("Retrieved data exceeds maxReceivedBytes");
+  }
+  std::istringstream stream(reply);
+  string line;
+  while (std::getline(stream, line)) {
+    lines.push_back(line);
+  }
+#endif
+  return lines;
+}
+
+void ZoneData::ZoneToCache(const RecZoneToCache::Config& config)
+{
+  if (config.d_sources.size() > 1) {
+    d_log->info("Multiple sources not yet supported, using first");
+  }
+  d_zone = DNSName(config.d_zone);
+  d_now = time(nullptr);
+
+  // We do not do validation, it will happen on-demand if an Indeterminate record is encountered when the caches are queried
+  // First scan all records collecting info about delegations ans sigs
+  // A this moment, we ignore NSEC and NSEC3 records. It is not clear to me yet under which conditions
+  // they could be entered in into the (neg)cache.
+
+  if (config.d_method == "axfr") {
+    d_log->info("Getting zone by AXFR");
+    getByAXFR(config);
+  }
+  else {
+    vector<string> lines;
+    if (config.d_method == "url") {
+      d_log->info("Getting zone by URL");
+      lines = getURL(config);
+    }
+    else if (config.d_method == "file") {
+      d_log->info("Getting zone from file");
+      lines = getLinesFromFile(config);
+    }
+    DNSResourceRecord drr;
+    ZoneParserTNG zpt(lines, d_zone);
+    zpt.setMaxGenerateSteps(0);
+
+    while (zpt.get(drr)) {
+      DNSRecord dr(drr);
+      parseDRForCache(dr);
+    }
+  }
+
```

Comment:
I know it's a bit cumbersome but it might be nice to check whether the configuration has been updated at this point, since the transfer/loading might have taken quite some time, and a different thread with a more recent configuration might have updated the cache with more recent values in the meantime.